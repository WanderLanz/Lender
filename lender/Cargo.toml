[package]
name = "lender"
authors = [
    "Lance Warden <wanderlanz@proton.me>",
    "Sebastiano Vigna <sebastiano.vigna@gmail.com>",
]
description = "A lending-iterator trait based on higher-rank trait bounds, with full std::iter::Iterator functionality"
version = "0.2.9"
edition = "2021"
repository = "https://github.com/WanderLanz/Lender"
license = "Apache-2.0 OR LGPL-2.1-or-later OR MIT"
readme = "README.md"
categories = ["rust-patterns", "no-std"]
keywords = ["lending", "iterator"]

[badges]
maintenance = { status = "experimental" }

[dependencies]
lender-derive = { path = "../lender-derive", version = "0.1.2", optional = true }
stable_try_trait_v2 = { path = "../stable_try_trait_v2", version = "1.75.1" }

[features]
default = ["derive"]
derive = ["lender-derive"]
