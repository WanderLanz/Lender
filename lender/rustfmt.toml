# These are just my preferences, feel free to change them to your liking.
max_width = 125
use_small_heuristics = "Max"
tab_spaces = 4
