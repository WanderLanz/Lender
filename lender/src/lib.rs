#![doc = include_str!("../README.md")]
#![no_std]

extern crate alloc;

#[cfg(doctest)]
#[allow(non_camel_case_types)]
/// ```rust,compile_fail
/// struct WindowsMut<'a, T> {
///     inner: &'a mut [T],
///     begin: usize,
///     len: usize,
/// }
/// impl<'a, T> Iterator for WindowsMut<'a, T> {
///     type Item = &mut [T];
///
///     fn next(&mut self) -> Option<Self::Item> {
///         let begin = self.begin;
///         self.begin = self.begin.saturating_add(1);
///         self.inner.get_mut(begin..begin + self.len)
///     }
/// }
/// ```
pub struct _Lender_Doctest_Sanity_Check;

#[cfg(feature = "derive")]
pub use lender_derive::for_;

#[doc(hidden)]
#[allow(private_bounds)]
pub trait ImplBound: ImplBoundPriv {}
#[doc(hidden)]
pub(crate) trait ImplBoundPriv {}
impl<T: ?Sized + ImplBoundPriv> ImplBound for T {}
#[doc(hidden)]
pub struct Ref<'lend, T: ?Sized>(&'lend T);
impl<T: ?Sized> ImplBoundPriv for Ref<'_, T> {}

mod adapters;
pub use adapters::*;
mod traits;
pub use traits::*;
pub mod higher_order;
mod sources;
pub use sources::*;
pub use stable_try_trait_v2 as try_trait_v2;

pub mod prelude {
    #[cfg(feature = "derive")]
    pub use lender_derive::for_;

    pub use crate::{
        from_into_iter, from_iter, hrc, hrc_mut, hrc_once, lend, DoubleEndedLender, ExactSizeLender, ExtendLender,
        FromLender, IntoIteratorExt, IntoLender, IteratorExt, Lend, Lender, Lending, WindowsMutExt,
    };
}
