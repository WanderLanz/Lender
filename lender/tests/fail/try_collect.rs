use crate::try_trait_v2::ChangeOutputType;
use lender::*;
#[derive(Debug)]
struct PrintOnDrop(String);

impl Drop for PrintOnDrop {
    fn drop(&mut self) {
        eprintln!("Drop {}", self.0)
    }
}

struct Wrapper<L>(L);

impl<L> FromLender<L> for Wrapper<L>
where
    L: IntoLender,
{
    fn from_lender(lender: L) -> Self {
        Self(lender)
    }
}

fn main() {
    let mut count = 0u8;
    let mut lender = lender::from_fn(
        (),
        hrc_mut!(for<'all> move |(): &'all mut ()| -> Option<Result<(), PrintOnDrop>> {
            let res = match count {
                0 => Some(Ok(())),
                count @ (2 | 4) => {
                    let err = PrintOnDrop(count.to_string());
                    Some(Err(err))
                }
                _ => None,
            };
            count += 1;
            res
        }),
    );
    let wrapper: ChangeOutputType<Result<(), _>, _> = lender.try_collect::<Wrapper<TryShunt<'_, _>>>();

    let mut try_shunt: TryShunt<_> = wrapper.expect("Expected to extract wrapper").0;
    while let Some(()) = try_shunt.next() {
        ()
    }
    while let Some(()) = try_shunt.next() {
        ()
    }
}
