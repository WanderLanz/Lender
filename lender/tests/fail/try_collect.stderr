error[E0597]: `lender` does not live long enough
  --> tests/fail/try_collect.rs:40:55
   |
25 |     let mut lender = lender::from_fn(
   |         ---------- binding `lender` declared here
...
40 |     let wrapper: ChangeOutputType<Result<(), _>, _> = lender.try_collect::<Wrapper<TryShunt<'_, _>>>();
   |                                                       ^^^^^^------------------------------------------
   |                                                       |
   |                                                       borrowed value does not live long enough
   |                                                       argument requires that `lender` is borrowed for `'static`
...
49 | }
   | - `lender` dropped here while still borrowed
   |
note: due to current limitations in the borrow checker, this implies a `'static` lifetime
  --> src/traits/lender.rs
   |
   |         for<'all> B: FromLender<TryShunt<'all, &'a mut Self>>,
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: implementation of `lender::FromLender` is not general enough
  --> tests/fail/try_collect.rs:40:62
   |
40 |     let wrapper: ChangeOutputType<Result<(), _>, _> = lender.try_collect::<Wrapper<TryShunt<'_, _>>>();
   |                                                              ^^^^^^^^^^^ implementation of `lender::FromLender` is not general enough
   |
   = note: `Wrapper<lender::TryShunt<'2, &mut lender::FromFn<(), {closure@$DIR/src/higher_order.rs:103:11: 103:39}>>>` must implement `lender::FromLender<lender::TryShunt<'1, &mut lender::FromFn<(), {closure@$DIR/src/higher_order.rs:103:11: 103:39}>>>`, for any lifetime `'1`...
   = note: ...but it actually implements `lender::FromLender<lender::TryShunt<'2, &mut lender::FromFn<(), {closure@$DIR/src/higher_order.rs:103:11: 103:39}>>>`, for some specific lifetime `'2`
